src/main/java/uk/gov/justice/hmpps/prison/service/digitalwarrant/DigitalWarrantService.java

```diff
@@ -0,0 +1,144 @@
+            .legalCaseType(legalCaseType)
+            .agencyLocation(agency)
+            .beginDate(courtCase.getBeginDate())
+            .offenderBooking(booking)
+            .caseSeq(sequence)
+            .build();
+
+        return offenderCourtCaseRepository.save(offenderCourtCase).getId();
+    }
+
+    @Transactional
+    public Long createOffenderOffence(Long bookingId, Offence offenderOffence) {
+        uk.gov.justice.hmpps.prison.repository.jpa.model.Offence offence = offenceRepository.findById((new PK(offenderOffence.getOffenceCode(), offenderOffence.getOffenceStatue()))).orElseThrow(EntityNotFoundException.withId(offenderOffence.getOffenceCode() + " " + offenderOffence.getOffenceStatue()));
+        OffenderCourtCase courtCase = offenderCourtCaseRepository.findById(offenderOffence.getCourtCaseId()).orElseThrow(EntityNotFoundException.withId(offenderOffence.getCourtCaseId()));
+        OffenderBooking booking = offenderBookingRepository.findByBookingId(bookingId).orElseThrow(EntityNotFoundException.withId(bookingId));
+        String offenceResultCode = offenderOffence.isGuilty() ? OffenceResultRepository.IMPRISONMENT : OffenceResultRepository.NOT_GUILTY;
+        OffenceResult result = offenceResultRepository.findById(offenceResultCode).orElseThrow(EntityNotFoundException.withId(offenceResultCode));
+
+        OffenderCharge offenderCharge = OffenderCharge.builder()
+            .offence(offence)
+            .dateOfOffence(offenderOffence.getOffenceDate())
+            .endDate(offenderOffence.getOffenceEndDate())
+            .offenderCourtCase(courtCase)
+            .offenderBooking(booking)
+            .resultCodeOne(result)
+            .mostSeriousFlag("N")
+            .build();
+
+        return offenderChargeRepository.save(offenderCharge).getId();
+    }
+
+    @Transactional
+    public Integer createOffenderSentence(Long bookingId, Sentence sentence) {
+        OffenderCourtCase courtCase = offenderCourtCaseRepository.findById(sentence.getCourtCaseId()).orElseThrow(EntityNotFoundException.withId(sentence.getCourtCaseId()));
+        OffenderBooking booking = offenderBookingRepository.findByBookingId(bookingId).orElseThrow(EntityNotFoundException.withId(bookingId));
+        OffenderCharge offenderCharge = offenderChargeRepository.findById(sentence.getOffenderChargeId()).orElseThrow(EntityNotFoundException.withId(sentence.getOffenderChargeId()));
+        SentenceCalcType sentenceCalcType = sentenceCalcTypeRepository.findById(new SentenceCalcType.PK(sentence.getSentenceType(), sentence.getSentenceCategory())).orElseThrow(EntityNotFoundException.withId(sentence.getSentenceType() + " " + sentence.getSentenceCategory()));
+
+        //TODO is this the correct way to set sequence?
+        Integer sequence = offenderSentenceRepository.findByOffenderBooking_BookingId(bookingId).stream().max(Comparator.comparing(OffenderSentence::getSequence)).map(os -> os.getSequence() + 1).orElse(1);
```

Comment:
I can't find the source code for this table so not sure if this is right or not. Would probably be quicker with a dedicated query for max sequence? Which would reduce the chance of a race condition actually happening - though it's probably already very very small. Anyway, worth revisiting post POC.

Reply:
Yea I think that would be optimised. This is working for now with an integration test. I wanted to just try a POC before digging into how its done in the NOMIS DB. Its very much work in progress!

I've got a ticket for improvements so I'll add this suggestion.